modules/core/src/matrix_expressions.cpp

```diff
@@ -1326,7 +1329,10 @@ void MatOp_AddEx::assign(const MatExpr& e, Mat& m, int _type) const
     else
     {
         e.a.convertTo(dst, e.a.type(), e.alpha);
-        cv::add(dst, e.s, dst);
+        if (e.a.channels() == 1)
+            cv::add(dst, e.s, dst);
+        else
+            cv::add(dst, e.s[0], dst);
```

Comment:
Both code branches are equal to:
```
cv::add(dst, e.s[0], dst);
```

so proposed condition doesn't make sense.

The real problem is how input "double" is stored through `Scalar`:
- `5` is `Scalar(5, 0, 0, 0)`
- or `5` is `Scalar:all(5)`
- also tests should properly handle `+ Scalar(5, 0, 0, 0)` and `Scalar:all(5)`

`.isReal()` check is not correct if we try to use `Scalar::all(5)` approach.
Probably we should raise exceptions on operations like `multichannel_mat + just scalar`.

Reply:
@alalek Thanks for review, to address your feedback:

- Yes, that if/else is redundant. I will just replace with `cv::add(dst, e.s, dst);`

- Raising exception in case of `mulitchannel_mat + just_scalar` is fine imo but I think it's better handled in a separate PR, where `isReal()` issue can be fully sorted out. That may take longer as the solution cannot be implemented inside matrix_expressions.cpp.

- As for this PR, I will change the tests above so that the `just scalar` is changed to the Scalar form: ex) `res = multiChannel_2x2__1_2_3 + Scalar(5, 0, 0);`, and then add the tests Scalar::all test you suggested.

This PR at least will fix OP's alpha blending issue and strengthen against regressions with the added tests. Is this fine?

Reply:
Patch is fine.
But it explodes another problem: `multiChannel + 5.0` result is not well defined.
Need to take a look on this in details.